torch/csrc/stable/library.h

```diff
@@ -168,3 +170,110 @@ class StableTorchLibraryInit final {
       __LINE__);                                                            \
   void STABLE_CONCATENATE(                                                  \
       STABLE_TORCH_LIBRARY_FRAGMENT_init_##ns##_, uid)(StableLibrary & m)
+
+namespace {
+
+template <class... T, std::size_t... I>
+std::tuple<T...> unbox_to_tuple_impl(
+    StableIValue* stack,
+    std::index_sequence<I...>) {
+  return std::make_tuple(to<T>(stack[I])...);
+}
+
+template <class... T>
+std::tuple<T...> unbox_to_tuple(StableIValue* stack) {
+  return unbox_to_tuple_impl<T...>(
+      stack, std::make_index_sequence<sizeof...(T)>());
+}
+
+template <class... T, std::size_t... I>
+void box_from_tuple_impl(
+    StableIValue* stack,
+    std::tuple<T...> vals,
+    std::index_sequence<I...>) {
+  ((stack[I] = from<T>(std::get<I>(vals))), ...);
+}
+
+template <class... T>
+void box_from_tuple(StableIValue* stack, std::tuple<T...> vals) {
+  box_from_tuple_impl<T...>(
+      stack, vals, std::make_index_sequence<sizeof...(T)>());
+}
+
+template <
+    typename ReturnType,
+    typename ParameterTypeList,
+    typename FuncT,
+    FuncT* func>
+struct boxer_impl {
+  static_assert(c10::guts::false_t<ReturnType>::value, "Unsupported");
+};
+
+template <
+    typename... ReturnTypes,
+    typename... ParameterTypes,
+    typename FuncT,
+    FuncT* func>
+struct boxer_impl<
+    std::tuple<ReturnTypes...>,
+    c10::guts::typelist::typelist<ParameterTypes...>,
+    FuncT,
+    func> {
+  void operator()(
+      StableIValue* stack,
+      uint64_t num_args,
+      uint64_t num_outputs) {
+    assert(num_args == sizeof...(ParameterTypes));
```

Comment:
TODO: this should become TORCH_INTERNAL_ASSERT (or whatever its header-only equivalent is)